### PR TITLE
Editor: Redux state behavior

### PR DESCRIPTION
**Pull request by aduth**
Related: #2994
Related: #3182

This pull request seeks to explore the second of two proposed mechanisms for managing post editor state in the Redux global state tree. The basis for this approach is that there is no state responsible for tracking the "current" edited post, and instead the assumption is that edits will always pass the desired site ID, post ID pairing to identify a post to be edited.

__Pseudo-code:__

```js
editPost( {
	title: 'Hello World'
}, 2916284, 417 );
```

__Advantages:__

- Action creators are more pure than in #3182
- No need to "start editing" a post before making revisions
 - Related: Other sections can more easily take advantage of these actions

__Disadvantages:__

- More burden on developer to determine site ID, post ID pairing in calling edit

__Open questions:__

~~Should we index edits by global ID as we do in the posts subtree? How would this work with (a) allowing multiple unsaved drafts to exist simultaneously (one-per-site?), and (b) editing posts which have yet to load (since global ID can only be determined for existing posts after post has loaded).~~
 - Decided not to use global ID

__Testing instructions:__

This has no effect on the Calypso application, because the editor reducer is not yet included in the global app state. Instead, ensure that Mocha tests pass by running `make test` from the project root directory or directly from `client/state/editor`.

### CODE REVIEW

**Comment by blowery** [comment]
One thing I like about this is that is naturally safe-guards actions dispatched asynchronously. That is, imagine
```
function addTempToTitle( post ) {
  fetchTemp().then( function( temp) {
    editPost( { title: post.title + "-" + temp }, post.ID, post.site_ID );
  } )
}
```

In the other proposal, where the post being edited is held in state, you have to jump through a hoop to make sure that the post currently being edited when fetchTemp's Promise resolves is still the post you started the process with. By passing along the ID and site_ID, you avoid that hoop.

**Comment by aduth** [comment]
This pull request is now ready for review. I've rebased to include missing JSDoc blocks and tests.

Since the editor reducer is not yet included in the global app state, there's no risk for breakage, so testing is a matter of reviewing the code and ensuring that Mocha tests pass.

File: client/state/editor/reducer.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ */
+export function posts( state = {}, action ) {
+	switch ( action.type ) {
+		case EDITOR_POST_EDIT:
+			const { post, siteId, postId = '' } = action;
+			state = Object.assign( {}, state, {
+				[ siteId ]: Object.assign( {}, state[ siteId ] )
+			} );
+
+			state[ siteId ][ postId ] = Object.assign( {}, state[ siteId ][ postId ], post );
```
Comment (gwwar) [inline client/state/editor/reducer.js]: Do we need cloneDeep here, or is the post object one level deep?

File: client/state/editor/test/reducer.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+		it( 'should track new post draft revisions by site ID', () => {
+			const state = posts( undefined, {
+				type: EDITOR_POST_EDIT,
+				siteId: 2916284,
+				post: { title: 'Ribs & Chicken' }
+			} );
+
+			expect( state ).to.eql( {
+				2916284: {
+					'': {
```
Comment (gwwar) [inline client/state/editor/test/reducer.js]: Is this what we call a dirty post currently? If we want to have more than one of these, we'll likely need to generate local ids.

**Comment by gwwar** [comment]
:+1: I like it, it keeps things simple even if we have the added burden of passing both a site and post id. I also don't think I see a clash with @artpi suggestion that we have smaller editor actions. One way of implementing that would be to create a number of simple wrappers:

```javascript
export function editPostTitle( title, siteId, postId ) {
    return editPost( { title: title } , siteId, postId );
}
```

File: client/state/editor/test/reducer.js (hunk 2 above)
Comment (aduth) [inline client/state/editor/test/reducer.js]: >Is this what we call a dirty post currently? If we want to have more than one of these, we'll likely need to generate local ids.

The empty string designates a new unsaved post draft. With this model, it's possible to have multiple simultaneous edits for any number of existing posts, but only a single new draft. I'd agree once we get to the point of locally saving the draft, the entity will be moved to `state.posts.items` with a temporary ID until the persisted ID is made available. For unsaved posts however, it seems logical to only allow a single draft at a time.

**Comment by aduth** [comment]
>One way of implementing that would be to create a number of simple wrappers:

Yeah, my understanding is this would be the desired implementation, still allowing `editPost` directly but offering more tailored wrappers which themselves call `editPost`.

**Comment by aduth** [comment]
Based on discussion yesterday, I wonder if it makes more sense for edited state to live at `posts.edits` and for the action creators to merged into `posts/actions.js`. This way we can still track edited state separately, while having an obvious connection with both state and action creators between edited state and the persisted post entities. Especially when we start to talk about saving post edits outside the context of the post editor (e.g. trashing a post on the posts list), the distinction between `state/posts/actions.js` and `state/editor/actions.js` becomes increasingly blurred.

/cc @mtias

File: client/state/editor/reducer.js (hunk 1 above)
Comment (aduth) [inline client/state/editor/reducer.js]: >Do we need cloneDeep here, or is the post object one level deep?
 
All but one property would be covered by a shallow merge: `discussion`. It's possible to save `discussion. comments_open` and `discussion.pings_open` separately, which would necessitate a deep merge. I'll explore options here and ensure that a test case is added.

Related docs: https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/posts/%24post_ID/

**Comment by mtias** [comment]
@aduth I think that may be a good path. We'll still likely have specific editor idiosyncrasies or functionality (thinking of things like collaborative editing) where we may want to keep the editor tree. But actions that modify posts, and the modified posts seem better suited for `state.posts`.

**Comment by aduth** [comment]
@mtias : [The rebased] 934032e moves all post edit state to the `posts` subtree. Decided against nesting it as its own subdirectory therein, since the reducer shouldn't require its own subkeys. It might be an idea to reconsider moving forward if it turns out that post editing grows to shadow the rest of general post-related state.

**Comment by aduth** [comment]
@gwwar : Pushed 6393f5d39e5b95c616913a513929c39531648c14 to address concerns raised at https://github.com/Automattic/wp-calypso/pull/3183#discussion_r53112843 .

File: client/state/posts/test/reducer.js [hunk 3]
```diff
@@ -468,4 +470,179 @@ describe( 'reducer', () => {
...
+			const state = edits( undefined, {} );
+
+			expect( state ).to.eql( {} );
+		} );
+
+		it( 'should track new post draft revisions by site ID', () => {
+			const state = edits( undefined, {
+				type: POST_EDIT,
+				siteId: 2916284,
+				post: { title: 'Ribs & Chicken' }
```
Comment (timmyc) [inline client/state/posts/test/reducer.js]: :meat_on_bone: never gets old

**Comment by timmyc** [comment]
Since this has no bearing on existing functionality, I say we merge this in before the branch gets too stale.  The logic looks sound, and the tests read well too.

**Pull request merged**